### PR TITLE
*: interpret timestamp datum based on session's time_zone

### PR DESCRIPTION
**Pull request by tiancaiamao**
1. FromPackedUint don't consider time_zone any more, the logic moved to Unflatten
2. use session's time_zone instead of time.Local
3. add a time.Location parameter to some Encode/Decode functions in tablecodec

After this change, timestamp value is interpreted based on session's time_zone variable:

```
mysql> create table t (ts timestamp);
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t values ('2017-04-27 22:40:42');
Query OK, 1 row affected (0.00 sec)

mysql> select * from t;
+---------------------+
| ts                  |
+---------------------+
| 2017-04-27 22:40:42 |
+---------------------+
1 row in set (0.01 sec)

mysql> set time_zone= '+10:00';
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t;
+---------------------+
| ts                  |
+---------------------+
| 2017-04-28 00:40:42 |
+---------------------+
1 row in set (0.01 sec)

mysql> set time_zone= '-6:00';
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t;
+---------------------+
| ts                  |
+---------------------+
| 2017-04-27 20:40:42 |
+---------------------+
1 row in set (0.00 sec)

```

Depends on https://github.com/pingcap/tidb/pull/3165

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @coocood @zimulala @shenli

File: table/tables/tables.go [hunk 1]
```diff
@@ -483,6 +484,14 @@ func (t *Table) RowWithCols(ctx context.Context, h int64, cols []*table.Column)
 	return v, nil
 }
 
+func getTimeZone(ctx context.Context) *time.Location {
```
Comment (coocood) [inline table/tables/tables.go]: This function is implemented multiple times, can we just implement it as a method of `SessionVars`?

Comment (coocood) [inline table/tables/tables.go]: Or just initialize it to `time.Local` when we create a new one.

File: store/tikv/mock-tikv/cop_handler_dag.go [hunk 2]
```diff
@@ -281,7 +283,8 @@ func (e *evalContext) decodeRelatedColumnVals(relatedColOffsets []int, handle in
 			}
 			continue
 		}
-		row[offset], err = tablecodec.DecodeColumnValue(value[offset], e.fieldTps[offset])
+		// TODO: Should use session's TimeZone instead of UTC.
```
Comment (coocood) [inline store/tikv/mock-tikv/cop_handler_dag.go]: There is a `TimeZoneOffset` field in `tipb.SelectRequest` and `tipb.DAGRequest`, we should use them.

**Comment by coocood** [comment]
There is a function `timeZoneOffset` in `executor/distsql.go` which is used to pass timezone to coprocessor. We should update it.

File: ddl/column.go [hunk 3]
```diff
@@ -347,7 +347,8 @@ func (d *ddl) backfillColumnInTxn(t table.Table, colMeta *columnMeta, handles []
 			return 0, errors.Trace(err)
 		}
 
-		rowColumns, err := tablecodec.DecodeRow(rowVal, colMeta.oldColMap)
+		// TODO: Should use session's TimeZone instead of UTC.
```
Comment (coocood) [inline ddl/column.go]: In DDL worker, there is no session.

File: inspectkv/inspectkv.go [hunk 4]
```diff
@@ -404,7 +405,8 @@ func rowWithCols(txn kv.Retriever, t table.Table, h int64, cols []*table.Column)
 		}
 		colTps[col.ID] = &col.FieldType
 	}
-	row, err := tablecodec.DecodeRow(value, colTps)
+	// TODO: Should use session's TimeZone instead of UTC.
```
Comment (coocood) [inline inspectkv/inspectkv.go]: I think we don't need to use session's TimeZone here.
Use UTC is ok.

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -305,6 +305,7 @@ type StatementContext struct {
 	IgnoreTruncate       bool
 	TruncateAsWarning    bool
 	InShowWarning        bool
+	TimeZone             *time.Location
```
Comment (coocood) [inline sessionctx/variable/session.go]: SessionVars already have a `TimeZone` field, this can be removed.

File: distsql/xeval/eval.go [hunk 6]
```diff
@@ -100,7 +100,7 @@ func (e *Evaluator) SetRowValue(handle int64, row [][]byte, relatedColIDs map[in
 		} else {
 			data := row[offset]
 			ft := e.fieldTps[offset]
-			datum, err := tablecodec.DecodeColumnValue(data, ft)
+			datum, err := tablecodec.DecodeColumnValue(data, ft, e.StatementCtx.TimeZone)
```
Comment (coocood) [inline distsql/xeval/eval.go]: Use time zone in `tipb.SelectReqest` instead.

File: ddl/column.go (hunk 3 above)
Comment (coocood) [inline ddl/column.go]: I think the best way is to convert the default value to UTC timezone in ddl API.
So we can just use UTC to decode row in DDL worker.

**Comment by coocood** [comment]
Please add tests

**Comment by shenli** [comment]
@coocood @BusyJay Should we also do this on coprocessor?

**Comment by shenli** [comment]
@tiancaiamao Please resolve the conflicts.

File: table/tables/tables.go (hunk 1 above)
Comment (tiancaiamao) [inline table/tables/tables.go]: Good idea.

File: store/tikv/mock-tikv/cop_handler_dag.go (hunk 2 above)
Comment (tiancaiamao) [inline store/tikv/mock-tikv/cop_handler_dag.go]: Why there is a TimeZone in the tipb protocol?

File: ddl/column.go (hunk 3 above)
Comment (tiancaiamao) [inline ddl/column.go]: Maybe, but I'm not quite sure.

**Comment by coocood** [comment]
@shenli
coprocessor already use the timezone we passed.

File: distsql/xeval/eval.go (hunk 6 above)
Comment (tiancaiamao) [inline distsql/xeval/eval.go]: No `tipb.SelectRequest` here

File: sessionctx/variable/session.go (hunk 5 above)
Comment (tiancaiamao) [inline sessionctx/variable/session.go]: Evaluator need this.
Evaluator just know StatementContext, it can't get TimeZone from SessionVars.

**Comment by tiancaiamao** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

File: executor/new_distsql.go [hunk 7]
```diff
@@ -172,7 +172,7 @@ func (e *IndexReaderExecutor) Next() (*Row, error) {
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		err = decodeRawValues(values, e.schema)
+		err = decodeRawValues(values, e.schema, e.ctx.GetSessionVars().GetTimeZone())
```
Comment (shenli) [inline executor/new_distsql.go]: You should also update IndexLoopupExecutor.

Comment (tiancaiamao) [inline executor/new_distsql.go]: IndexLookupExecutor getRow is in Raw format, it will be decode somewhere else.

**Comment by shenli** [comment]
LGTM

**Pull request merged**